src/Compiler/test/Microsoft.NET.Sdk.Razor.SourceGenerators.Tests/RazorSourceGeneratorTests.TagHelpers.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Microsoft.NET.Sdk.Razor.SourceGenerators;
+
+public partial class RazorSourceGeneratorTests
+{
+    [Fact]
+    public async Task CustomTagHelper()
+    {
+        // Arrange
+        var project = CreateTestProject(new()
+        {
+            ["Views/Home/Index.cshtml"] = """
+                @addTagHelper *, TestProject
+
+                <email>
+                    custom tag helper
+                    <email>nested tag helper</email>
+                </email>
+                """
+        }, new()
+        {
+            ["EmailTagHelper.cs"] = """
+                using Microsoft.AspNetCore.Razor.TagHelpers;
+
+                public class EmailTagHelper : TagHelper
+                {
+                    public override void Process(TagHelperContext context, TagHelperOutput output)
+                    {
+                        output.TagName = "a";
+                    }
+                }
+                """
+        });
+        var compilation = await project.GetCompilationAsync();
+        var driver = await GetDriverAsync(project);
+
+        // Act
+        var result = RunGenerator(compilation!, ref driver);
+
+        // Assert
+        Assert.Contains("EmailTagHelper", result.GeneratedSources.Single().SourceText.ToString());
```

Comment:
Instead of generating noisy baselines, perhaps asserts like these would be enough (at least for most tests)? This one verifies that the `EmailTagHelper` is instantiated. If it wasn't discovered, the source would instead just include `WriteLiteral("<email>...")`. (Which makes me realize I should also verify that.)

Reply:
Definitely in favor of less noisy baseline files. Trying to decide what the best balance is here. The rule we have in Roslyn used to be a Neal-ism that went something along the lines of: 

> The correctness of the compiler is judged by what the code outputs 

That is why so many of our tests in Roslyn just very what the code prints. It _proves_ the compiler did the correct thing. 

Is there a morale equivalent of that for tag helpers? Is it for example looking at the final HTML here only?

Reply:
> Is there a morale equivalent of that for tag helpers? Is it for example looking at the final HTML here only?

Yes, "executing" the Razor template and verifying the HTML output would be awesome - but it requires the ASP.NET runtime (`RenderTreeBuilder` implementation) - which we don't have in Razor repo. But I imagine we could have some simplified implementation or perhaps use ASP.NET NuGet (seems it wouldn't matter if it's an old one).

Reply:
I prototyped "rendering tests" here: https://github.com/jjonescz/razor/compare/9b615b753c1fb66b483d09badbcb2ba2721e6946...rendering-tests - can render and verify HTML output of Razor Pages and Blazor components.